### PR TITLE
[ioctl] Build block bucketlist command line into new ioctl

### LINKED ISSUE #3266: [ioctl] build block bucketlist command line into new ioctl
### What would you like to be added
 - Step 1: create `bcbucketlist.go` and `bcbucketlist_test.go` files in `./ioctl/newcmd/bc`. 
 - Step 2: refactor  command  https://github.com/iotexproject/iotex-core/blob/master/ioctl/cmd/bc/bcbucketlist.go in `bcbucketlist.go`
 - Step 3: add unit test in xx_test.go

#### Notes:
1. Use [client interface](https://github.com/iotexproject/iotex-core/blob/master/ioctl/client.go) to construct the Cobra command.
2. [Output package](https://github.com/iotexproject/iotex-core/blob/master/ioctl/output) is deprecated. `e.g.` `output.PrintError` could be replaced with `errors.Wrap`.
3. Global variables (config.ReadConfig  config.DefaultConfigFile etc.) have been wrapped into client interface. Please use [client interface](https://github.com/iotexproject/iotex-core/blob/master/ioctl/client.go) to access them if needed
4. The logic of new code should be similar to the old one.
5. In each unit test, the results should be checked whether they are as the expected. 

#### Example:
1. https://github.com/iotexproject/iotex-core/blob/master/ioctl/newcmd/account/accountinfo.go
2. https://github.com/iotexproject/iotex-core/blob/master/ioctl/newcmd/account/accountdelete.go
3. https://github.com/iotexproject/iotex-core/blob/master/ioctl/newcmd/account/accountcreateadd.go

### Why is this needed
Unit tests couldn't be added for the old ioctl command (./ioctl/cmd)

### PR DESCRIPTION
**Pull request by LuckyPigeon**
# Description

Build `bcbucketlist` command in new ioctl, with the following note.

* Use [client interface](https://github.com/iotexproject/iotex-core/blob/master/ioctl/client.go) to construct the Cobra command.
* Output package is deprecated, replace it with errors package.
* Replace `fmt.Println` with `cmd.Println`
* Global variables (config.ReadConfig config.DefaultConfigFile etc.) have been wrapped into client interface. Please use [client interface](https://github.com/iotexproject/iotex-core/blob/master/ioctl/client.go) to access them if needed
* Refactor unit test to cover the modification.
* Check results in each unit test.

Fixes #3266 

## Type of change
Please delete options that are not relevant.
- [x] Code refactor or improvement

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [x] make test
- [x] TestNewBCBucketListCmd

**Test Configuration**:
- Firmware version: Ubuntu 21.04
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: ioctl/newcmd/bc/bcbucketlist.go [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+
+	"github.com/iotexproject/iotex-core/ioctl"
+	"github.com/iotexproject/iotex-core/ioctl/config"
+	"github.com/iotexproject/iotex-core/ioctl/util"
+)
+
+var (
+	_validMethods = []string{
```
Comment (huof6829) [inline ioctl/newcmd/bc/bcbucketlist.go:25]: move to line33

File: ioctl/newcmd/bc/bcbucketlist.go [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+		Use:   use,
+		Short: short,
+		Long:  long,
+		Args:  cobra.MinimumNArgs(2),
+		Example: `ioctl bc bucketlist voter [VOTER_ADDRESS] [OFFSET] [LIMIT]
+	ioctl bc bucketlist cand [CANDIDATE_NAME] [OFFSET] [LIMIT]`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			var err error
+
+			cmd.SilenceUsage = true
```
Comment (huof6829) [inline ioctl/newcmd/bc/bcbucketlist.go:65]: move to line 63

File: ioctl/newcmd/bc/bcbucketlist.go [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+		for _, bucket := range m.Bucketlist {
+			lines = append(lines, bucket.String())
+		}
+	}
+	return strings.Join(lines, "\n")
+}
+
+// getBucketList get bucket list from chain by voter address
+func getBucketListByVoter(client ioctl.Client, addr string, offset, limit uint32) error {
+	address, err := util.GetAddress(addr)
```
Comment (huof6829) [inline ioctl/newcmd/bc/bcbucketlist.go:113]: use client.AddressWithDefaultIfNotExist

File: ioctl/newcmd/bc/bcbucketlist.go [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+	if err != nil {
+		return errors.Wrap(err, "")
```
Comment (huof6829) [inline ioctl/newcmd/bc/bcbucketlist.go:115]: return err

File: ioctl/newcmd/bc/bcbucketlist.go [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+	return &cobra.Command{
+		Use:   use,
+		Short: short,
+		Long:  long,
+		Args:  cobra.MinimumNArgs(2),
+		Example: `ioctl bc bucketlist voter [VOTER_ADDRESS] [OFFSET] [LIMIT]
+	ioctl bc bucketlist cand [CANDIDATE_NAME] [OFFSET] [LIMIT]`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			var err error
```
Comment (huof6829) [inline ioctl/newcmd/bc/bcbucketlist.go:63]: move line69

File: ioctl/newcmd/bc/bc.go [hunk 6]
```diff
@@ -129,3 +131,50 @@ func GetProbationList(client ioctl.Client, epochNum uint64) (*iotexapi.ReadState
...
+	readStakingDataRequest *iotexapi.ReadStakingDataRequest,
+) (*iotextypes.VoteBucketList, error) {
+	apiServiceClient, err := client.APIServiceClient()
+	if err != nil {
+		return nil, err
+	}
+	method := &iotexapi.ReadStakingDataMethod{Method: methodName}
+	methodData, err := proto.Marshal(method)
+	if err != nil {
+		return nil, output.NewError(output.SerializationError, "failed to marshal read staking data method", err)
```
Comment (huof6829) [inline ioctl/newcmd/bc/bc.go:148]: replace output with errors.Wrap

File: ioctl/newcmd/bc/bc.go [hunk 7]
```diff
@@ -129,3 +131,50 @@ func GetProbationList(client ioctl.Client, epochNum uint64) (*iotexapi.ReadState
...
+	}
+	requestData, err := proto.Marshal(readStakingDataRequest)
+	if err != nil {
+		return nil, output.NewError(output.SerializationError, "failed to marshal read staking data request", err)
```
Comment (huof6829) [inline ioctl/newcmd/bc/bc.go:152]: replace output with errors.Wrap

File: ioctl/newcmd/bc/bc.go [hunk 8]
```diff
@@ -129,3 +131,50 @@ func GetProbationList(client ioctl.Client, epochNum uint64) (*iotexapi.ReadState
...
+		ctx = metautils.NiceMD(jwtMD).ToOutgoing(ctx)
+	}
+
+	response, err := apiServiceClient.ReadState(ctx, request)
+	if err != nil {
+		sta, ok := status.FromError(err)
+		if ok {
+			return nil, output.NewError(output.APIError, sta.Message(), nil)
+		}
+		return nil, output.NewError(output.NetworkError, "failed to invoke ReadState api", err)
```
Comment (huof6829) [inline ioctl/newcmd/bc/bc.go:173]: the same

File: ioctl/newcmd/bc/bc.go [hunk 9]
```diff
@@ -129,3 +131,50 @@ func GetProbationList(client ioctl.Client, epochNum uint64) (*iotexapi.ReadState
...
+	jwtMD, err := util.JwtAuth()
+	if err == nil {
+		ctx = metautils.NiceMD(jwtMD).ToOutgoing(ctx)
+	}
+
+	response, err := apiServiceClient.ReadState(ctx, request)
+	if err != nil {
+		sta, ok := status.FromError(err)
+		if ok {
+			return nil, output.NewError(output.APIError, sta.Message(), nil)
```
Comment (huof6829) [inline ioctl/newcmd/bc/bc.go:171]: the same

File: ioctl/newcmd/bc/bc.go [hunk 10]
```diff
@@ -129,3 +131,50 @@ func GetProbationList(client ioctl.Client, epochNum uint64) (*iotexapi.ReadState
...
+		}
+		return nil, output.NewError(output.NetworkError, "failed to invoke ReadState api", err)
+	}
+	bucketlist := iotextypes.VoteBucketList{}
+	if err := proto.Unmarshal(response.Data, &bucketlist); err != nil {
+		return nil, output.NewError(output.SerializationError, "failed to unmarshal response", err)
```
Comment (huof6829) [inline ioctl/newcmd/bc/bc.go:177]: the same

File: ioctl/newcmd/bc/bcbucketlist.go [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+
+// getBucketList get bucket list from chain by voter address
+func getBucketList(client ioctl.Client, method string, addr string, offset, limit uint32) error {
+	var (
+		bl      *iotextypes.VoteBucketList
+		address string
+		err     error
+	)
+
+	if method == "voter" {
```
Comment (huof6829) [inline ioctl/newcmd/bc/bcbucketlist.go:115]: check method twice by line 79 and 115.

File: ioctl/newcmd/bc/bcbucketlist.go [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+				}
+			}
+			if len(s) > 1 {
+				limit, err = strconv.ParseUint(s[1], 10, 64)
+				if err != nil {
+					return errors.Wrap(err, "invalid limit")
+				}
+			}
+			switch method {
+			case "voter":
```
Comment (huof6829) [inline ioctl/newcmd/bc/bcbucketlist.go:85]: use const variable for "voter"、"cand"  can be better

File: ioctl/newcmd/bc/bcbucketlist.go [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+				if err != nil {
+					return err
+				}
+				bucketlist = append(bucketlist, bucket)
+			}
+			message := bucketlistMessage{
+				Node:       client.Config().Endpoint,
+				Bucketlist: bucketlist,
+			}
+			fmt.Println(message.String())
```
Comment (huof6829) [inline ioctl/newcmd/bc/bcbucketlist.go:111]: cmd.Println

File: ioctl/newcmd/bc/bcbucketlist.go [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+			case "cand":
+				bl, err = getBucketListByCandidateName(client, addr, uint32(offset), uint32(limit))
+			default:
+				return errors.New("unknown <method>")
+			}
+			if err != nil {
+				return err
+			}
+			var bucketlist []*bucket
+			for _, b := range bl.Buckets {
```
Comment (huof6829) [inline ioctl/newcmd/bc/bcbucketlist.go:100]: use `for range` twice in line 100 and line 127,  they can be combined once. delete bucketlistMessage and handle it directly

File: ioctl/newcmd/bc/bcbucketlist.go [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+				Node:       client.Config().Endpoint,
+				Bucketlist: bucketlist,
+			}
+			fmt.Println(message.String())
+			return nil
+		},
+	}
+}
+
+type bucketlistMessage struct {
```
Comment (huof6829) [inline ioctl/newcmd/bc/bcbucketlist.go:117]: delete

**Comment by huof6829** [comment]
add test for `GetBucketList`

File: ioctl/newcmd/bc/bc_test.go [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+)
+
+func TestGetBucketList(t *testing.T) {
+	require := require.New(t)
+	ctrl := gomock.NewController(t)
+	client := mock_ioctlclient.NewMockClient(ctrl)
+	apiServiceClient := mock_iotexapi.NewMockAPIServiceClient(ctrl)
+	request := &iotexapi.ReadStakingDataRequest{}
+	response := &iotexapi.ReadStateResponse{}
+
```
Comment (huof6829) [inline ioctl/newcmd/bc/bc_test.go:29]: delete spaceline

File: ioctl/newcmd/bc/bc_test.go [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+	client.EXPECT().APIServiceClient().Return(apiServiceClient, nil).Times(2)
+
+	t.Run("get bucket list", func(t *testing.T) {
+		expectedValue := &iotextypes.VoteBucketList{
+			Buckets: []*iotextypes.VoteBucket{},
+		}
+
```
Comment (huof6829) [inline ioctl/newcmd/bc/bc_test.go:36]: delete spaceline

File: ioctl/newcmd/bc/bc_test.go [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+		apiServiceClient.EXPECT().ReadState(gomock.Any(), gomock.Any()).Return(response, nil)
+
+		result, err := GetBucketList(client, iotexapi.ReadStakingDataMethod_BUCKETS_BY_CANDIDATE, request)
+		require.NoError(err)
+		require.Contains(result.String(), expectedValue.String())
+	})
+
+	t.Run("failed to invoke ReadState api", func(t *testing.T) {
+		expectedErr := errors.New("failed to invoke ReadState api")
+
```
Comment (huof6829) [inline ioctl/newcmd/bc/bc_test.go:46]: delete spaceline

**Comment by LuckyPigeon** [comment]
@huof6829 Thanks for the fix!

**Comment by huof6829** [comment]
> @huof6829 Thanks for the fix!

I add more test datas to cover more codes.

**Review by huof6829 — APPROVED** [review]

**Review by Liuhaai — APPROVED** [review]

**Pull request merged**